Support for CI Testing using Azure Pipelines/Docker

keithhealy opened the pull request:
# Summary

- This PR is a feature
- It adds support for CI testing of Axom using Azure Pipelines/Docker.
- Currently enables testing on Linux, Windows, and Mac.
- Linux testing uses following compilers: Clang 4,5,6 and GCC 6,7,8
- Dockerfiles located in axom/scripts/docker were used to generate docker images stored on Docker Hub.  See README.

gzagaris commented on azure-pipelines.yml:
```diff
@@ -13,38 +13,54 @@ strategy:
-      Compiler_ImageName: 'axom/compilers:gcc-6'
-      CMAKE_EXTRA_FLAGS: '-DAXOM_ENABLE_SIDRE=Off'
+      Compiler_ImageName: 'axom/tpls:gcc-6'
+      CMAKE_EXTRA_FLAGS: '-DAXOM_ENABLE_SIDRE=ON -DBUILD_SHARED_LIBS_ON'
```
You don't have to explicitly set `AXOM_ENABLE_SIDRE=ON` here. It's `ON" by default. Also, there is a typo in `-DBUILD_SHARED_LIBS_ON`. I think the intent probably was `-DBUILD_SHARED_LIBS=ON`, but we shouldn't need this.

keithhealy commented:
I have removed these flags.

On Mon, Jan 27, 2020 at 8:54 PM George Zagaris <notifications@github.com>
wrote:

> *@gzagaris* commented on this pull request.
> ------------------------------
>
> In azure-pipelines.yml
> <https://github.com/LLNL/axom/pull/155#discussion_r371609852>:
>
> > @@ -13,38 +13,54 @@ strategy:
>    matrix:
>      linux_gcc6:
>        VM_ImageName: 'ubuntu-16.04'
> -      Compiler_ImageName: 'axom/compilers:gcc-6'
> -      CMAKE_EXTRA_FLAGS: '-DAXOM_ENABLE_SIDRE=Off'
> +      Compiler_ImageName: 'axom/tpls:gcc-6'
> +      CMAKE_EXTRA_FLAGS: '-DAXOM_ENABLE_SIDRE=ON -DBUILD_SHARED_LIBS_ON'
>
> You don't have to explicitly set AXOM_ENABLE_SIDRE=ON here. It's ON" by
> default. Also, there is a typo in -DBUILD_SHARED_LIBS_ON. I think the
> intent probably was -DBUILD_SHARED_LIBS=ON`, but we shouldn't need this.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/LLNL/axom/pull/155?email_source=notifications&email_token=AMALA6NVTOVUSC4HSSSKXGTQ763BNA5CNFSM4KMHTO72YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCTHU2IA#pullrequestreview-349130016>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMALA6L4T32YGUAMNELA2ELQ763BNANCNFSM4KMHTO7Q>
> .
>

kennyweiss commented on scripts/docker/README:
```diff
@@ -0,0 +1,37 @@
+FROM axom/compilers:<compiler>
+
+The remaining dockerfile commands attempt to build the Axom third party libraries using the specified compiler.
+The following commands update the compiler and add Fortran compiler support:
```
@davidbeckingsale -- should we fold these commands into the compiler docker images that way we do not need to rerun them each time?

(or am I reading this wrong?)

kennyweiss commented on scripts/docker/README:
```diff
@@ -0,0 +1,37 @@
+
+Then the Axom and RAJA source is retrieved and RAJA is built and installed: (WORKDIR functions as a 'cd' command)
+RUN git clone --recursive https://github.com/LLNL/axom.git
+RUN git clone --recursive https://github.com/LLNL/RAJA.git
```
I think we should be able to build ``RAJA`` directly from uberenv.  

Let's take care of that in a separate PR, after this is merged.

agcapps replied:
I agree.

kennyweiss commented on scripts/docker/README:
```diff
@@ -0,0 +1,37 @@
+Dockerfiles are used to create Docker images.  Docker images are the 'recipes' used to build virtual environments called containers.
```
For readability on github, it might be better to convert this to a markdown file.
(we can take care of this separately).

This is what it looks like now on github: https://github.com/keithhealy/axom/tree/develop/scripts/docker

keithhealy replied:
I changed the file to markdown format.

kennyweiss replied:
Thanks @keithhealy -- the writeup is really helpful.

Could you please rename it to ``README.md`` ?

keithhealy replied:
Changed to README.md and fixed some formatting errors.

kennyweiss commented on scripts/azure-pipelines/linux-build_and_test.sh:
```diff
@@ -16,21 +18,40 @@ function or_die () {
+./config-build.py -hc /home/axom/axom/host-configs/docker/${HOST_CONFIG}.cmake -DENABLE_GTEST_DEATH_TESTS=ON
+
+or_die cd build-$HOST_CONFIG-debug
+or_die make -j 3
```
Do you know how many threads we can use on azure pipelines?

It looks like we're building with 3 threads and runing tests with 8.

keithhealy replied:
I think we can run up to 10 parallel jobs.
I don't know how many threads of operation there are in a job.
https://docs.microsoft.com/en-us/azure/devops/pipelines/licensing/concurrent-jobs?view=azure-devops

kennyweiss replied:
Thanks -- in that case, should we increase this to ``j10``?

kennyweiss commented on scripts/azure-pipelines/linux-build_and_test.sh:
```diff
@@ -16,21 +18,40 @@ function or_die () {
+echo HOST_CONFIG
+echo $HOST_CONFIG
+
+./config-build.py -hc /home/axom/axom/host-configs/docker/${HOST_CONFIG}.cmake -DENABLE_GTEST_DEATH_TESTS=ON
```
> -DENABLE_GTEST_DEATH_TESTS=ON

Please remove this cmake option -- it is set in the individual ``host-configs``

kennyweiss commented on scripts/docker/README:
```diff
@@ -0,0 +1,37 @@
+Finally, the uberenv script is invoked to build the libraries and store them in the axom_tpls directory.  (The host config files
+produced during this process had been added to the Axom repo under "axom/host-configs/docker")
+WORKDIR "/home/axom/axom"
+RUN ./scripts/uberenv/uberenv.py --spack-config-dir=./scripts/uberenv/spack_configs/docker --spec=%clang@4.0.0^mpich@3.0.4 --prefix=/home/axom/axom_tpls -k
```
@keithhealy -- can you please add a line about requiring ``mpich`` for docker.  This will document the trouble you had with the default spack MPI.

kennyweiss commented on scripts/docker/dockerfile_clang-4:
```diff
@@ -0,0 +1,30 @@
+RUN cmake ../. -DENABLE_TESTS=OFF
+RUN make 
+RUN sudo make install
+
```
This probably doesn't matter since we're going to fix the RAJA build problem, 
but should we delete the ``raja/build`` directory?

kennyweiss commented on host-configs/docker/docker-linux-ubuntu16.04-x86_64-gcc@8.1.0.cmake:
```diff
@@ -0,0 +1,94 @@
+# Other machine specifics
+##############
+
+set(ENABLE_GTEST_DEATH_TESTS ON CACHE BOOL "")
```
Note: The ``ENABLE_GTEST_DEATH_TESTS`` option is being set in the host-configs, so we can remove it from the ``linux-build_and_test`` shell script.

kennyweiss commented on scripts/docker/dockerfile_clang-4:
```diff
@@ -0,0 +1,30 @@
+FROM axom/compilers:clang-4
```
Thanks for adding the dockerfiles @keithhealy !

kennyweiss commented on scripts/azure-pipelines/linux-build_and_test.sh:
```diff
@@ -16,21 +18,40 @@ function or_die () {
+find ./axom/sidre -type f -exec chmod 644 {} \;
+find . -name "*.xml"
+
+#set -x
```
@keithhealy -- should we fold the ``DO_BUILD``, ``DO_TEST`` logic back into the above script and delete the stuff below?

keithhealy replied:
Sorry I didn't clean up this file very well.
The find/chmod commands were necessary because some of the sidre files didn't have their permissions set correctly.  The 'find *.xml' is for my own debugging and I'll remove it along with the commented out code.

keithhealy replied:
As for the DO_BUILD and DO_TEST, I assumed there wouldn't be a case where we didn't want to do both (build and test) so I removed the logic.  I can restore it if that is preferable.

Again I apologize for not doing a better job cleaning up this file.

kennyweiss replied:
Not a problem.  I agree that this seems unnecessary.

@white238, or others, do we need ``DO_BUILD`` and ``DO_TEST`` for our linux CI builds?
Perhaps we had it there to support our coverage tests? (which we'd like to add back in the near future).

rhornung67 replied:
I think we need those since we will be able to do build tests for some compilers but may not be able to run tests; e.g., CUDA.

keithhealy replied:
Restored the DO_BUILD and DO_TEST logic.

kennyweiss commented on scripts/azure-pipelines/linux-build_and_test.sh:
```diff
@@ -16,21 +18,40 @@ function or_die () {
+
+
+find ./axom/sidre -type d -exec chmod 755 {} \;
+find ./axom/sidre -type f -exec chmod 644 {} \;
```
@keithhealy -- are these ``chmod``s necessary? Or was it part of your testing?

keithhealy replied:
Sorry I commented in the wrong place.  The chmod commands were necessary.  The find "*.xml" is not and I'll remove it.

kennyweiss commented on scripts/docker/README:
```diff
@@ -0,0 +1,37 @@
+Once completed, this container can then be used to build and test the Axom application on a Docker-using CI web tool such as Azure Pipelines.
+Already built Docker images containing these third party libraries can be accessed on Docker Hub at this location:
+https://hub.docker.com/r/axom/tpls/tags
+
```
@keithhealy -- what is the process for updating these TPLs on dockerhub?  Could you please document this?

keithhealy replied:
I'll add to the README.

kennyweiss commented on scripts/azure-pipelines/linux-build_and_test.sh:
```diff
@@ -16,21 +18,40 @@ function or_die () {
+
+or_die cd build-$HOST_CONFIG-debug
+or_die make -j 3
+make CTEST_OUTPUT_ON_FAILURE=1 test ARGS='-T Test -VV -j8'
```
Please add ``or_die`` at the beginning of this line.

kennyweiss commented on scripts/azure-pipelines/linux-build_and_test.sh:
```diff
@@ -16,21 +18,40 @@ function or_die () {
+
+find ./axom/sidre -type d -exec chmod 755 {} \;
+find ./axom/sidre -type f -exec chmod 644 {} \;
+find . -name "*.xml"
```
Should this also be wrapped with ``or_die``?

keithhealy replied:
If the permissions change fails some number of sidre tests will fail but the whole run would not be affected.  Adding or_die would cause these test results to be lost, I believe, as the whole job would be marked as failed.  I should test this.

rhornung67 commented:
@kennyweiss I agree. I didn't mean for those changes to be in this PR. I just don't think we should spend time testing compilers that folks don't use.

kennyweiss commented on azure-pipelines.yml:
```diff
@@ -57,7 +73,13 @@ steps:
+  displayName: 'Visual Studio Build'
 - script: |
     cd build
     make
```
Note for future -- it looks like we're building on windows, but not yet testing.  We should fix that in a separate PR.

Tag: @agcapps

agcapps replied:
Tag noted.

keithhealy commented on scripts/docker/README:
```diff
@@ -0,0 +1,37 @@
+Dockerfiles are used to create Docker images.  Docker images are the 'recipes' used to build virtual environments called containers.
```
I have added to the README to discuss creating images and uploading to Docker Hub.
Regarding the failed tests.  It's a little confusing but it looks like the results you are seeing with the failures was produced by an automatically triggered test run that uses the merged pull request changes. You can see these results here:
https://dev.azure.com/axom/axom/_build/results?buildId=839&view=ms.vss-test-web.build-test-results-tab
Whereas the test results of the unmerged pull request are here:
https://dev.azure.com/healy22/Axom%20CI/_build/results?buildId=206&view=ms.vss-test-web.build-test-results-tab

kennyweiss replied:
Thanks @keithhealy that's interesting -- one in is axom's azure space and the other is in your azure space.

I think we want to indicate failure  if either of these has test failures -- premerge or postmerge, since the latter will break tests for others.

Also, could you please merge develop into your branch (probably after fixing the test failure reporting), since I think it is out of date?

keithhealy commented on scripts/uberenv/spack_configs/docker/compilers.yaml:
```diff
@@ -0,0 +1,80 @@
+    modules: []
+    operating_system: ubuntu16.04
+    paths:
+      cc:  /usr/bin/gcc
```
Yes.  Each Docker image has one version installed.

cyrush commented on azure-pipelines.yml:
```diff
@@ -69,8 +91,10 @@ steps:
-    docker run --rm --user='root' -v `pwd`:/home/axom $(Compiler_ImageName) chown -R axom /home/axom
-    docker run --rm  -v `pwd`:/home/axom -e TEST_TARGET -e COMPILER -e DO_BUILD -e DO_TEST -e CMAKE_EXTRA_FLAGS $(Compiler_ImageName) ./scripts/azure-pipelines/linux-build_and_test.sh
+    echo " -e $TEST_TARGET -e $COMPILER -e $DO_BUILD -e $DO_TEST -e $CMAKE_EXTRA_FLAGS $(Compiler_ImageName) ./scripts/azure-pipelines/linux-build_and_test.sh"
+    docker run --rm --user='root' -v `pwd`:/home/axom/axom $(Compiler_ImageName) chown -R axom /home/axom
```
You can actually run the docker container directly with the azure job spec, here is an example:

```
  - job: Ubuntu_Cuda
    pool:
      vmImage: 'ubuntu-16.04'
    container: alpinedav/ascent-ci:cuda-9.2-dev
```

Not sure if that really simplifies things or not in this case, but I wanted to mention it, since I saw we are calling docker ourselves to launch the container.

gzagaris commented on azure-pipelines.yml:
```diff
@@ -13,38 +13,54 @@ strategy:
       TEST_TARGET: 'osx_gcc'
+    windows:
+      VM_ImageName: 'windows-2019'
+      CMAKE_EXTRA_FLAGS: '-DAXOM_ENABLE_SIDRE=Off'
```
Is there a particular reason to turn off Sidre on OS X and windows?

kennyweiss replied:
@gzagaris -- I think the issue is that we're not building TPLs on these platforms and ``sidre`` requires ``conduit``.  

We'll have to fix this in a separate PR.  
* @agcapps is working on the Windows part.
* For OS X -- does anyone know if docker/azure supports this?

gzagaris commented on scripts/docker/dockerfile_clang-6:
```diff
@@ -0,0 +1,28 @@
+
+RUN git clone --recursive https://github.com/LLNL/axom.git
+
+RUN git clone --recursive https://github.com/LLNL/RAJA.git
```
Is the RAJA clone and build step needed? Doesn't the call to `uberenv` below (line 24) install RAJA? (Ditto for all the other docker files.)

kennyweiss replied:
Keith had trouble getting this to work through docker.  

The plan is to fix this is a separate PR, after this one has merged.

kennyweiss commented:
#### RE: docker in the host-config name.

@gzagaris -- I am pretty sure this is the name that uberenv generates.
In the short term, it is probably best to keep things as they are so we can just copy them without renaming. Longer term, perhaps we can modify uberenv to omit the ``docker-`` prefix in the generated host-config.

kennyweiss commented:
At this point, the only unresolved issue is that the ``sidre_lulesh`` tests are timing out on the gcc docker configurations, which I think is due to the interaction between MPI and OpenMP on a constrained azure resource (our node only has 10 threads, and the sidre_lulesh task is being called on 8 ranks with 2 threads each.  

Oddly enough:
* the slam_lulesh test is being called with 8 ranks and 4 threads each and is passing on the gcc configs
* the same sidre_lulesh tests are passing on the clang configs

See: https://dev.azure.com/axom/axom/_build/results?buildId=877&view=ms.vss-test-web.build-test-results-tab

Since this isn't currently showing up as a test failure, I think we should merge the current PR and resolve this separately.  And also figure out how to get azure to indicate a "check" failure when some of the tests are failing.

Any objections to merging as is?

kennyweiss commented:
I created a ticket to track the open issues and suggested improvements for our CI plan: https://github.com/LLNL/axom/issues/162

Please add to it if I missed anything.

gzagaris commented:
Thanks for creating the ticket @kennyweiss -- I agree we should merge this as is now and fine tune things in a subsequent PRs.

kennyweiss commented:
Thanks @keithhealy!

keithhealy commented:
A couple of comments.
1. Uberenv adds the hostname to the beginning of the host config file name when it is created.  When you run an instance of a Docker image the hostname that is assigned is a random 12 digit alpha-numeric string.  I removed these and added 'Docker' because it was easier to remember (and wouldn't require changing the test scripts each time).  Removing it should not be a problem. 
2. Azure does recognize that the sidre-lulesh tests fail and reports this fact in the published results.  It does not, however, mark the CI jobs as failed when there are failed tests. The rationale seems to be that the jobs have succeeded (in running the tests and publishing the results) regardless of the contents of those results. I believe there are two options to remedy this.
   A. Fail the 'Build and Test' job when tests fail.  Then set the 'Publish Test Results' job to always run.  Normally a job failing will prevent subsequent jobs from running.  In this case we want the test results to be published even if the previous job fails due to failed tests.  Otherwise we get a notification of a failure but no record of which tests failed other than what's in the log.
  B. Fail the 'Publish Test Results' job when tests fail.  This is probably the easiest solution even if it is slightly misleading.  That is, the job succeeded in publishing the test results but has been marked as failed due to a failed result.
Let me know what you think.

gzagaris commented:
@keithhealy , if there is a test failure, we'd want to make sure that merging of the branch is not going to be allowed. Would this be the case with what you propose in 2.B?

kennyweiss commented:
@keithhealy -- I agree with @gzagaris -- We want "check" jobs to fail when one or more unit tests fail, or at the very least, we'd want the github "check" reason to indicate that there were failed tests.  Right now, the only way to see that a test failed is to click on the link that says "All tests passed" and click through a "Tests" tab.

Do we have flexibility over the name "Publish Test Results" ?  Or can we add a task after that named something like "Verify Test Results" ?

keithhealy commented:
There is a way to define checks that must be passed before a pull request can be merged. This is located under Settings/branches of the repo.  I believe it should be possible to check for failing CI tasks.  I'll investigate.
As for flexibility over the tasks itself, the existing ones have a lot of optional settings.  For example, the PublishTestResults task has an option called 'failTaskOnFailedTests' which will fail the task if any tests fail.  For the most flexibility we could just add a script task which finds and parses the results file.  I don't think we could rename this though.

kennyweiss commented:
> For example, the PublishTestResults task has an option called 'failTaskOnFailedTests' which will fail the task if any tests fail. 

Thanks @keithhealy -- That sounds like the one we want.

cyrush commented:
It seems the docker run is not returning the resulting return code?

It may simplify things to have azure use the docker container to run the job, instead of running docker ourselves if there are are issues with it not propagating the return codes. 

Some tests not failing is only one example, what about when a compile fails?

Note - for each script block, azure pipelines will only check the return code of the last executed task. This caught me off guard a bit back, so I wanted to mention it.

In my experience -- A ctest azure step should stop if any tests fail.  (ctest will return non-zero)

In fact if you look at Ascent,  we wanted to make sure to post test results even when ctest failed.  To do this, we had to use a `condition: always()` specifier to the PublishTestResults.



```
      - script: |
           #################################
           # test
           #################################
           # find spack installed cmake
           export ROOT_DIR=`pwd`
           export CMAKE_BIN_DIR=`ls -d ${ROOT_DIR}/uberenv_libs/spack/opt/spack/*/*/cmake*/bin`
           export PATH=${CMAKE_BIN_DIR}:$PATH
           which ctest
           cd build
           # run ctest
           ctest -T test --output-on-failure -V
        displayName: 'Run Unit Tests'
      - task: PublishTestResults@2
        inputs:
          testResultsFormat: 'cTest'
          testResultsFiles: '**/Test.xml'
        displayName: 'Unit Test Results'
        condition: always()
```

gzagaris commented:
Yes -- having an indicator to easily know which actual stage has failed (build or test) is useful. 

I took a look at the ascent setup. I like how it gives a summary of each of the jobs(or configurations) in the pull request. That way we can know whether the Windows or OS X build/ or tests have failed in the Pull Request page, without having to click on the "Details" to go to the azure-pipelines page.  Would that be easy to setup for Axom?

cyrush commented:
It should be straight forward, you just use top level stages + jobs.

Another nice thing to fall out of this is that you don't need a bunch of optional steps. The per platform steps are simply written for each platform specific job.

keithhealy commented:
The compile steps are run through the 'or_die' function which will cause the script to exit with a non-zero return code if the compile fails.  The 'make test' step is not run through 'or_die' because I wanted to make sure the permissions was set correctly on the test results and that the 'PublisTestResults' task was always run.
As Cyrus mentioned, setting the 'always()' condition on PublishTestResults is one way to fix this.  Another is to set the PublishTestResults task to fail if any tests fail.  

You can add checks that must be passed when merging by going to Settings/Branches for the repo and selecting 'Require status checks to pass before merging'.  Next click 'Add Rule' and click the checks you want to require before a merge is allowed.  You should get something like this (Scroll down to the merge pane):
https://github.com/keithhealy/MyAxom/pull/1

keithhealy commented:
Sorry, that should be "going to Settings/Branches for the repo and selecting 'Add Rule' under Branch Protection Rules"

gzagaris commented:
Thanks -- I think doing the same for Axom would be useful.
